File: cognitodemo/authenticate.js
const Cognito = require("@aws-sdk/client-cognito-identity-provider");
const jwt = require("aws-jwt-verify");
const crypto = require("crypto");

const userPoolId = ""; // Obtain from the AWS console
const clientId = "";  // Obtain from the AWS console
const clientSecret = "";  // Obtain from the AWS console
const username = "theuser";
const password = "supersecret";
const email = "your email address goes here";

function secretHash(clientId, clientSecret, username) {
  const hasher = crypto.createHmac('sha256', clientSecret);
  hasher.update(`${username}${clientId}`);
  return hasher.digest('base64');
}


const accessVerifier = jwt.CognitoJwtVerifier.create({
  userPoolId: userPoolId,
  tokenUse: "access",
  clientId: clientId,
});

const idVerifier = jwt.CognitoJwtVerifier.create({
  userPoolId: userPoolId,
  tokenUse: "id",
  clientId: clientId,
});

async function main() {
  const client = new Cognito.CognitoIdentityProviderClient({
    region: "ap-southeast-2",
  });

  console.log("Getting auth token");

  // Get authentication tokens from the Cognito API using username and password
  const command = new Cognito.InitiateAuthCommand({
    AuthFlow: Cognito.AuthFlowType.USER_PASSWORD_AUTH,
    AuthParameters: {
      USERNAME: username,
      PASSWORD: password,
      SECRET_HASH: secretHash(clientId, clientSecret, username),
    },
    ClientId: clientId,
    
  });

  res = await client.send(command);
  console.log(res);

  // ID Tokens are used to authenticate users to your application
  const IdToken = res.AuthenticationResult.IdToken;
  const IdTokenVerifyResult = await idVerifier.verify(IdToken);
  console.log(IdTokenVerifyResult);

  // Access tokens are used to link IAM roles to identities for accessing AWS services
  // Most students will not use these
  const accessToken = res.AuthenticationResult.AccessToken;
  const accessTokenVerifyResult = await accessVerifier.verify(accessToken);
  console.log(accessTokenVerifyResult);
}

main();

File: cognito_python/authenticate.py
import boto3
import hmac, hashlib, base64 

username= "testuser"
password= "MySuperSecret99!"
email= "Your email goes here"
client_id = ""  # Obtain from the AWS console
client_secret = ""  # Obtain from the AWS console

def secretHash(clientId, clientSecret, username):
    message = bytes(username + clientId,'utf-8') 
    key = bytes(clientSecret,'utf-8') 
    return base64.b64encode(hmac.new(key, message, digestmod=hashlib.sha256).digest()).decode() 


def authenticate():
    client = boto3.client("cognito-idp", region_name="ap-southeast-2")
    try:
        response = client.initiate_auth(
            AuthFlow="USER_PASSWORD_AUTH",
            AuthParameters={
                "USERNAME": username,
                "PASSWORD": password,
                "SECRET_HASH": secretHash(client_id, client_secret, username)
            },
            ClientId=client_id  # match signUp.py
        )
        tokens = response["AuthenticationResult"]
        # Optionally verify tokens here using jose or cognito public keys
        return tokens
    except Exception as e:
        print(f"Error during authentication: {e}")
        return None

authenticate_response = authenticate()
print("Authentication successful:", authenticate_response)
